### PR TITLE
[UR] Add lifetime validation to validation layer

### PR DESCRIPTION
**Pull request by kswiecicki**
... and refactor the validation layer mako template.

Relies on: https://github.com/oneapi-src/unified-runtime/pull/1029.
Solves: https://github.com/oneapi-src/unified-runtime/issues/867.

### CODE REVIEW

File: scripts/templates/helper.py [hunk 1]
```diff
@@ -1297,45 +1297,79 @@ def get_loader_epilogue(namespace, tags, obj, meta):
 
     return epilogue
 
+
+def get_event_wait_list_functions(specs, namespace, tags):
+    funcs = []
+    for s in specs:
+        for obj in s['objects']:
+            if re.match(r"function", obj['type']):
+                if any(x['name'] == 'phEventWaitList' for x in obj['params']) and any(
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply scripts/templates/helper.py:1306]: Well, I'm not sure what's the idea behind this function, I've just refactored and moved the `_get_create_get_retain_release_functions` function.

**Review by pbalcer — COMMENTED** [review]
right now this checks whether the provided ptr is a valid ptr to something, not necessarily an object of the expected type. Shouldn't be too hard to add?

File: scripts/core/INTRO.rst [hunk 2]
```diff
@@ -256,6 +256,8 @@ Layers currently included with the runtime are as follows:
      - Enables non-adapter-specific parameter validation (e.g. checking for null values).
    * - UR_LAYER_LEAK_CHECKING
      - Performs some leak checking for API calls involving object creation/destruction.
+   * - UR_LAYER_LIFETIME_VALIDATION
+     - Performs lifetime validation on objects used in API calls. Requires UR_LAYER_LEAK_CHECKING.
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply scripts/core/INTRO.rst:260]: Yes, it's not a problem. I was considering either enabling it automatically or making it a requirement.
I'll change it.

File: scripts/templates/valddi.cpp.mako [hunk 3]
```diff
@@ -72,39 +74,52 @@ namespace ur_validation_layer
 
         }
 
+        if( context.enableLifetimeValidation ) {
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply scripts/templates/valddi.cpp.mako]: Done.

File: scripts/core/INTRO.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (kswiecicki) [reply scripts/core/INTRO.rst]: I've added an explanation, take a look.

**Comment by kswiecicki** [comment]
> right now this checks whether the provided ptr is a valid ptr to something, not necessarily an object of the expected type. Shouldn't be too hard to add?

I've added `std::type_index` info to the `RefRuntimeInfo` and check for the index equality when checking a reference validity.
I was trying to also print handle type names but they were mangled and showed internal type names like `ur_adapter_handle_t_ *` instead of `ur_adapter_handle_t`.

File: test/layers/validation/fixtures.hpp [hunk 4]
```diff
@@ -30,36 +30,57 @@ struct urTest : ::testing::Test {
...
     }
 
     std::vector<ur_adapter_handle_t> adapters;
+};
+
+struct valAdapterTest : valAdaptersTest {
+
+    void SetUp() {
+        valAdaptersTest::SetUp();
+        adapter = adapters[0]; // TODO - which to choose?
```
Comment (pbalcer) [inline test/layers/validation/fixtures.hpp:60]: the way the tests are run there will only ever be one adapter. so [0] is fine.

**Review by pbalcer — APPROVED** [review]

**Comment by pbalcer** [comment]
> I was trying to also print handle type names but they were mangled and showed internal type names like `ur_adapter_handle_t_ *` instead of `ur_adapter_handle_t`.

I think that's fine. If you want to do this, one approach might be to pass the type name as a string from the generated layer to the class with runtime info. Would allow you to do away with typeindex and get correct type names. Not sure if it's worth it.

**Review by pbalcer — APPROVED** [review]

File: scripts/templates/valddi.cpp.mako [hunk 5]
```diff
@@ -74,37 +76,35 @@ namespace ur_validation_layer
...
+        %if func_name in tp_handle_funcs['create']:
         if( context.enableLeakChecking && result == UR_RESULT_SUCCESS )
         {
-            refCountContext.incrementRefCount(${object_param}, true);
+            refCountContext.createRefCount(*${tp['name']});
         }
-        %elif func_name == n + "AdapterGet":
-        if( context.enableLeakChecking && phAdapters && result == UR_RESULT_SUCCESS )
+        %elif func_name in tp_handle_funcs['get']:
+        if( context.enableLeakChecking && ${tp['name']} && result == UR_RESULT_SUCCESS )
```
Comment (PatKamin) [inline scripts/templates/valddi.cpp.mako:102]: Can the `name` be an empty field?

Comment (PatKamin) [inline scripts/templates/valddi.cpp.mako:102]: Is it possible for `tp['name']` to be empty?

File: test/layers/validation/lifetime.cpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// Copyright (C) 2023 Intel Corporation
```
Comment (PatKamin) [inline test/layers/validation/lifetime.cpp:1]: Update dates in all files.

File: test/layers/validation/lifetime.cpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "fixtures.hpp"
+
+TEST_F(urTest, testUrAdapterHandleLifetimeExpectFail) {
+    size_t size = 0;
+    ur_adapter_handle_t adapter = (ur_adapter_handle_t)0xC0FFEE;
+    ur_adapter_info_t info_type = UR_ADAPTER_INFO_BACKEND;
+    ASSERT_EQ(urAdapterGetInfo(adapter, info_type, 0, nullptr, &size),
+              UR_RESULT_ERROR_INVALID_ARGUMENT);
```
Comment (PatKamin) [inline test/layers/validation/lifetime.cpp]: You return the actual `size` here. Add an assert for this var or ignore it by passing a `nullptr` and removing this variable. Same below.

File: cmake/match.py [hunk 8]
```diff
@@ -5,77 +5,131 @@
...
+        # match or process tags
+        if re.fullmatch(pattern, input_line):
+            input_idx += 1
+            match_idx += 1
+            tags_in_effect = []
+        elif Tag.OPT in tags_in_effect:
+            match_idx += 1
+            tags_in_effect.remove(Tag.OPT)
+        elif Tag.IGNORE in tags_in_effect:
+            ignored_lines.append(input_line + "\n")
```
Comment (lukaszstolarczuk) [inline cmake/match.py:126]: `\n` -> `os.linesep`...?

File: cmake/match.py [hunk 9]
```diff
@@ -5,77 +5,131 @@
...
+            input_idx += 1
```
Comment (lukaszstolarczuk) [inline cmake/match.py:127]: hmm.. when is IGNORE tag removed from `tags_in_effect`...?

File: cmake/match.py [hunk 10]
```diff
@@ -5,77 +5,131 @@
...
 
-# check if all lines match in a file
-# lines in a match file can contain regex inside of double curly braces {{}}
+# Check if all input file content matches match file content.
+# Lines in a match file can contain regex inside of double curly braces {{}}.
+# Regex patterns are limited to single line.
+#
+# List of available special tags:
+# {{OPT}}    - makes content in the same line as the tag optional
+# {{IGNORE}} - ignores all content until next match or end of input file
```
Comment (lukaszstolarczuk) [inline cmake/match.py:14]: just confirming - by "next match" you mean "next IGNORE tag"? perhaps you could be more specific, please

File: scripts/templates/valddi.cpp.mako (hunk 5 above)
Comment (kswiecicki) [reply scripts/templates/valddi.cpp.mako:102]: It should never be an empty field unless specification was incorrectly formulated.

Comment (kswiecicki) [reply scripts/templates/valddi.cpp.mako:102]: Answered above.

File: test/layers/validation/fixtures.hpp [hunk 11]
```diff
@@ -30,36 +30,57 @@ struct urTest : ::testing::Test {
...
         }
         urTest::TearDown();
     }
 
     std::vector<ur_adapter_handle_t> adapters;
+};
+
+struct valAdapterTest : valAdaptersTest {
+
+    void SetUp() {
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply test/layers/validation/fixtures.hpp]: Done.

File: test/layers/validation/fixtures.hpp [hunk 12]
```diff
@@ -30,36 +30,57 @@ struct urTest : ::testing::Test {
...
+
+    void SetUp() override {
+        valAdaptersTest::SetUp();
+
+        uint32_t count;
+        ASSERT_EQ(urPlatformGet(adapters.data(),
+                                static_cast<uint32_t>(adapters.size()), 0,
+                                nullptr, &count),
+                  UR_RESULT_SUCCESS);
+        ASSERT_NE(count, 0);
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply test/layers/validation/fixtures.hpp]: Done.

File: test/layers/validation/lifetime.cpp (hunk 6 above)
Comment (kswiecicki) [reply test/layers/validation/lifetime.cpp]: Done.

File: test/layers/validation/lifetime.cpp (hunk 7 above)
Comment (kswiecicki) [reply test/layers/validation/lifetime.cpp]: I've removed the `size` param.

File: cmake/match.py (hunk 8 above)
Comment (kswiecicki) [reply cmake/match.py]: Done.

File: cmake/match.py (hunk 10 above)
Comment (kswiecicki) [reply cmake/match.py]: By next match I mean the next line that's successfully matched, next `{{IGNORE}}` tag will do nothing since `{{IGNORE}}` tag is already in effect.

Comment (lukaszstolarczuk) [reply cmake/match.py]: och, I see, then please write that in the comment 😉 `the next successfully matched line`

btw, I'd write: `or the end of the input ...` (missing "the" x2)

File: test/layers/validation/lifetime.cpp (hunk 7 above)
Comment (kswiecicki) [reply test/layers/validation/lifetime.cpp]: Actually, I need `size` param for this function to fail at the expected staged. I don't check its value because I don't want to  rely on the null adapter generic implementation for `urDeviceGetInfo` which does nothing.

File: scripts/templates/valddi.cpp.mako (hunk 5 above)
Comment (PatKamin) [reply scripts/templates/valddi.cpp.mako:102]: This check `tp['name']` can be removed then.

Comment (kswiecicki) [reply scripts/templates/valddi.cpp.mako:102]: This is not a check, I'm retrieving the value at `tp['name']` with `${...}`.

**Review by PatKamin — APPROVED** [review]

File: scripts/core/common.yml [hunk 13]
```diff
@@ -102,7 +102,7 @@ name: "$x_kernel_handle_t"
 --- #--------------------------------------------------------------------------
 type: handle
 desc: "Handle of a queue object"
-class: $xPlatform
+class: $xQueue
```
[the remark this answers is not in the document]
Comment (kswiecicki) [reply scripts/core/common.yml:105]: Done.

Comment (pbalcer) [inline scripts/core/common.yml:105]: this needs to be moved into another PR. Otherwise this one lgtm and we can merge it.

Comment (kswiecicki) [reply scripts/core/common.yml]: Done. The queue fix is here https://github.com/oneapi-src/unified-runtime/pull/1285.
